app/core/rate_limiting/gemini_limiter.py

```
@@ -0,0 +1,335 @@
+                f"rpm_used={result.metadata.rpm_used}/{result.metadata.rpm_limit}, "
+                f"rpd_used={result.metadata.rpd_used}/{result.metadata.rpd_limit}"
+            )
+            
+            return result
+            
+        except Exception as e:
+            self.logger.error(f"Rate limit check failed for {model}: {e}")
+            # Fail safe - block request
+            raise GeminiServiceUnavailableException(
+                f"Rate limiting service unavailable: {e}"
+            )
```

Ruff (0.11.9)

Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)